src/transformers/trainer.py

```diff
@@ -2565,7 +2565,7 @@ def _prepare_input(self, data: Union[torch.Tensor, Any]) -> Union[torch.Tensor,
             return type(data)(self._prepare_input(v) for v in data)
         elif isinstance(data, torch.Tensor):
             kwargs = {"device": self.args.device}
-            if self.deepspeed and data.dtype != torch.int64:
+            if self.deepspeed and data.dtype not in (torch.int32, torch.int64):
```

Comment:
Should we check other types of ints here to have a more complete fix?

Reply:
@sgugger , my initial fix was to check whether first layer is embedding. it would cover all such cases - [prev commit ](https://github.com/huggingface/transformers/pull/21800/commits/efc7e7f3a5c7df783ba82411652a8f0a30a360e0#diff-ed55888e6665791fe92cc8fc0c499da54f4ace6738551cd9a2591881cda076deL2568)
We could get back to it.

or would you like to check whether input is any int ?

Reply:
I'd prefer to check all kinds of ints here (int16, int8, uint8)

Reply:
@sgugger, done

Reply:
Won't this be a problem with quantized int8 weights?